File: Practice Set 1/Dynamic Programming/DP50 - Minimum Cost to Cut the Stick.py
def recursive():
    """
        Time complexity is exponential and space complexity is O(n).
    """

    def solve(rod, i, j):
        if i > j:
            return 0
        min_cost = 1e6
        for k in range(i, j + 1):
            cost = rod[j + 1] - rod[i - 1] + solve(rod, i, k - 1) + solve(rod, k + 1, j)
            min_cost = min(min_cost, cost)
        return min_cost

    def get_min_cost_to_cut(cut_coordinates, length_of_rod):
        cut_coordinates.sort()
        rod = [0,] + cut_coordinates + [length_of_rod,]
        return solve(rod, 1, len(cut_coordinates))

    print(get_min_cost_to_cut([1, 3, 4, 5], 7))
    print(get_min_cost_to_cut([1, 3], 4))
    print(get_min_cost_to_cut([1, 3, 4], 5))
    print(get_min_cost_to_cut([1, 3, 4, 7], 10))
    print(get_min_cost_to_cut([1, 3], 10))


recursive()

File: Java/DynamicProgramming/DP50 - Minimum Cost to Cut the Stick.java
package DynamicProgramming;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

class DP50Recursive {
    private static Integer solve(List<Integer> arr, Integer i, Integer j) {
        if (i > j) {
            return 0;
        }
        int minCost = Integer.MAX_VALUE;
        for (int k = i; k <= j; k += 1) {
            int cost = arr.get(j + 1) - arr.get(i - 1) + solve(arr, i, k - 1) + solve(arr, k + 1, j);
            minCost = Math.min(cost, minCost);
        }
        return minCost;
    }

    public static Integer getMinCostToCutRod(List<Integer> arr, Integer length) {
        int n = arr.size();
        arr.sort(null);
        List<Integer> rod = Stream.of(List.of(0), arr, List.of(length)).flatMap(List::stream).toList();
        return DP50Recursive.solve(rod, 1, n);
    }
}

class Solution {
    public static void main(String[] args) {
        System.out.println("Recursive Solution");
        System.out.println(DP50Recursive.getMinCostToCutRod(Arrays.asList(1, 3, 4, 5), 7));
        System.out.println(DP50Recursive.getMinCostToCutRod(Arrays.asList(1, 3), 4));
        System.out.println(DP50Recursive.getMinCostToCutRod(Arrays.asList(1, 3, 4), 5));
        System.out.println(DP50Recursive.getMinCostToCutRod(Arrays.asList(1, 3, 4, 7), 10));
        System.out.println(DP50Recursive.getMinCostToCutRod(Arrays.asList(1, 3), 10));
        System.out.println(DP50Recursive.getMinCostToCutRod(Arrays.asList(5, 6, 1, 4, 2), 9));
    }
}